### ISSUE TITLE
featurize_dataframe inplace is extremely slow for featurizers w/ lots (1k+) features

### ISSUE THREAD

**Issue by ardunn** [hackingmaterials/matminer#368]
### problem
In featurize_dataframe, there is this section of code:

```python
        if inplace:
            # Update the existing dataframe
            for k in labels:
                df[k] = res[k]
            return df
        else:
            # Create new dataframe and ensure columns are ordered properly
            new = pd.concat([df, res], axis=1)
            return new[df.columns.tolist() + res.columns.tolist()]
```

When inplace is True, pandas goes about appending a new column for each feature. This is a problem for featurizers which return lots of features, because internally in pandas some lists are being appended to for each feature - appending to native Python lists is amortized O(1) (approximately O(n) non-amortized, but observed significantly worse), so these operations can be surprisingly slow (hours, or **much** longer than the featurization time!).

Changing the inplace code to `df[labels] = res[labels]` has the same problem.

## solution
The inplace=False option may also be amortized O(1), but it is significantly faster. For example, I featurized ~1.1k samples with 60k features:

```
inplace = True
elapsed time (s): 538.517

inplace = False
elapsed time (s): 3.16
```
Note that these times are *just for assigning the columns to the df, not featurization*. Featurization itself took less than 10s. 

The inplace = False time can be lowered further by not reordering the columns:
```python
        #else:
           # Create new dataframe and ensure columns are ordered properly
            #new = pd.concat([df, res], axis=1)
            #return new[df.columns.tolist() + res.columns.tolist()]
        else:
            return pd.concat([df, res], axis=1)
```

```
inplace = False
elapsed time (s): 0.58
```

For some featurizers, this inplace operation is prohibitively expensive, even for small datasets. In the extreme, you have something like BagofBonds which can generate 150k+ features on a 1.1k sample dataset in 30s, but will take _literally 12 hours_ to assign them to a dataframe inplace. Enabling inplace=False assigns to the df in 0.8s. 

## proposed change
For these speed reasons, I suggest we make the default behavior of `featurize_dataframe` inplace=False, cleaning up the code a bit in the process.

**Comment by WardLT** [comment]
The proposed changes sound great! Do you already have a break with these changes? 

I'm totally fine with `inplace=False` as a default

**Comment by ardunn** [comment]
yeah unfortunately it breaks a lot of stuff, see #371

**Comment by ardunn** [comment]
closed via #383
